### PR TITLE
RDRP-519: Numeric PG for SAS extract

### PR DESCRIPTION
**Pull request by JenCheshire**
- `check_pg_mapper.py` created to check for any inconsistencies across the PG mappers.  I have kept it in for the peer review, but it can be removed before merging.

- Long forms with numeric PG: new column "pg_numeric" created with values at staging. Column 201 is then converted to alpha, and the column name remains 201.

- The `sic_to_pg_mapper()` now takes the formtype as list.  The default is set to short forms (`0006`), but we need to be able to apply the function to imputed long forms at output

- Short forms: created an additional mapper sic_pg_num.csv from the file SIC_to_PG_UTF-8.csv, use colums SIC 2007_CODE and 2016 > Form PG; de-duplicate on SIC and check that it's a many-to-one
Stage and validate sic_pg_num.csv, use validation.validate_many_to_one

- In GB SAS output, apply the sic_pg_num to get the numerical PG for short forms and imputed long forms, and put it in pg_numeric column.   The non-imputed long form codes are kept in pg_numeric.  Also applied to the TAU output which was also using the mock mapper.

- Change the gb sas and tau schema so that the selgrpno is taken from pg_numeric column. 

Closes RDRP-519

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: check_pg_mappers.py
Comment (zorge69) [inline check_pg_mappers.py:1]: Good to include this checking script. I would add de-duplication on the "many" SICside before the comparison. Still, it's not essential for live production, so happy to pass.

File: src/imputation/tmi_imputation.py
Comment (zorge69) [inline src/imputation/tmi_imputation.py:1]: I am not sure - why does it have to be changed to a list? How is it related to PG?

File: src/outputs/outputs_main.py [hunk 1]
```diff
@@ -28,9 +28,8 @@ def run_outputs(
     cora_mapper: pd.DataFrame,
     postcode_mapper: pd.DataFrame,
     itl_mapper: pd.DataFrame,
-    pg_alpha_num: pd.DataFrame,
     pg_num_alpha: pd.DataFrame,
-    sic_pg_alpha: pd.DataFrame,
```
Comment (zorge69) [inline src/outputs/outputs_main.py:33]: sic_pg_alpha is still needed. We need to for short forms that are not imputed.

File: src/imputation/tmi_imputation.py
Comment (JenCheshire) [reply src/imputation/tmi_imputation.py:1]: Thanks @zorge69 I though a list might be easiest way to specify which form/s the `sic_to_pg_mapper()` should be run on.  It is used a few times in the pipeline, but on different forms, so I’ve used .isin() in the function.  e.g. `run_pg_conversion()` in staging it is only used on short forms, here in tmi_imputation it is only used on long forms, and in the outputs, it is run on both.

File: src/outputs/outputs_main.py (hunk 1 above)
Comment (JenCheshire) [reply src/outputs/outputs_main.py:33]: I think that's being run in staging?

**Review by zorge69 — APPROVED** [review]
The code is looking good, but I was not able to run it because of my git issue

**Pull request merged**